src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,219 @@
+    {
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->configuration['table_name'], 'm')
+                ->where('m.delivered_at is null')
+                ->andWhere('m.available_at <= :now')
+                ->andWhere('m.queue = :queue')
+                ->orderBy('available_at', 'ASC')
+                ->setMaxResults(1);
+
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL(),
+                [
+                    ':now' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+                    'queue' => $this->configuration['queue'],
+                ]
+            )->fetch();
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+            $sql = $this->driverConnection->createQueryBuilder()
+                ->update($this->configuration['table_name'])
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id');
+            $this->executeQuery($sql, [
+                ':id' => $doctrineEnvelope['id'],
+                ':delivered_at' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+            ]);
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Throwable $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+    }
+
+    public function nack($id): bool
+    {
+        return $this->driverConnection->update($this->configuration['table_name'], ['delivered_at' => null], ['id' => $id]) > 0;
+    }
+
+    public function requeueMessages()
+    {
+        $redeliverLimit = new \DateTime('now');
+        $redeliverLimit->modify(sprintf('-%d second', $this->configuration['redeliver_timeout']));
+        $sql = $this->driverConnection->createQueryBuilder()
+            ->update($this->configuration['table_name'])
+            ->set('delivered_at', ':delivered_at')
+            ->andWhere('delivered_at < :redeliver_limit')
+            ->andWhere('queue = :queue')
+            ->getSQL();
+
+        $this->executeQuery($sql, [
+            ':delivered_at' => null,
+            ':queue' => $this->configuration['queue'],
+            ':redeliver_limit' => $redeliverLimit->format('Y-m-d H:i:s'),
+        ]);
+    }
+
+    private function executeQuery(string $sql, array $parameters = [])
+    {
+        $stmt = $this->driverConnection->prepare($sql);
```

Comment:
While trying locally, this statement is throwing the `TableNotFoundException` (with an SQLite adapter), it should be in the `try`. In order to be able to re-execute after setup, just re-run this prepare line as well.